src/marshmallow/fields.py

```diff
@@ -1616,6 +1620,79 @@ def __init__(self, *args, **kwargs):
         self.validators.insert(0, validator)
 
 
+class IP(Field):
+    """A IP address field.
+
+    :param bool exploded: If `True`, serialize ipv6 address in long form, ie. with groups
+        consisting entirely of zeros included."""
+
+    default_error_messages = {"invalid_ip": "Not a valid IP address."}
+
+    def __init__(self, *args, exploded=False, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.exploded = exploded
+
+    def _serialize(self, value, attr, obj, **kwargs) -> typing.Optional[str]:
+        if value is None:
+            return None
+        if self.exploded:
+            return value.exploded
+        return value.compressed
+
+    def _deserialize(
+        self, value, attr, data, **kwargs
+    ) -> typing.Optional[typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address]]:
+        if value is None:
+            return None
+        try:
+            if isinstance(value, (int, bytes)):
```

Comment:
Fields that inherit from `String` don't have to store their deserialized value internally as a string. Because fields are both dumpers and loaders, it is better to think of `String` as `String -> Any -> String`. Subclassing `String` allows you to specialize `Any` to a more concrete type.

IMO explicitly checking for ints and bytes is unnecessary. If you inherit form `String`, `super()._deserialize()` will provide all the type validation that is needed. `ipaddress` takes care of validating the format.

Reply:
`ipaddress` utilities are flexible in the terms of input ip addresses representations. Beside classic string representations it also accepts binary and integer representations. Explicit type checking was my arbitrary choice in order to narrow down the original `ipaddress` functionality. In my opinion deserialization of following JSON
```
{"ip": 1}
```
into this object representation
```
{"ip": IPv4Address('0.0.0.1')}
```
may be unexpected. But as I can see it leads to confusion on the implementation level, thus I have added constructor parameter determining whether decimal representations should be accepted. The parameter is `False` by default.

Reply:
Only passing strings to `ipaddress` seems reasonable. Excluding a subset of the types that `ipaddress` accepts seems fragile. If `ipaddress` starts accepting new types in the future this would let other types through. Allowing only strings would make the type checking explicit and resilient to upstream changes. The `String` field is a nice way to inherit this behavior declaratively.

Reply:
Accepting only strings when deserializing IP fields is something that I can agree on. But achieving this via inheriting from `String` is something that is not so clear for me. From the sources and the discussion I can see two approaches on the `String` base class.

First approach is the one I described in previous comment, where `String` based fields deserialize into `str`, and have specific validator attached. Examples of such fields are `URL` and `Email` which are just string values with additional format constraints. If so, `UUID` filed nor `IP` field should inherit from `String`, since they are not based on a validator and doesn't serialize into string.

The second approach, is the one where `String` based fields are like `String -> Any -> String`. So, if I understand it correctly, all fields that serialize to strings, regardless of deserialized type, should inherit from `String`. Then, `UUID` and `IP` fields should inherit from `String`, as well as all `DateTime` and `Time` fields, which currently inherit from `Field`.

Reply:
Since there is no feedback on this matter, I decided on an approach that better suits my understanding of the structure of the module.

Reply:
The only thing `String._deserialize()` would do differently is decode bytes into a UTF-8 string using `utils.ensure_text_type()`. That has been the default behavior for all of the string fields for a long time, but it's not clear what the use case was from the original commit.

@sloria Do you remember why byte strings are assumed to be UTF-8 encoded and decoded into strings? Should we do that for IP address strings also?

Original commit: https://github.com/marshmallow-code/marshmallow/commit/ff18e782be9d04557f027116b147f16bcecb9999

Reply:
I suppose that it may be related to the fact that `Schema` objects can have `render_module` altered to some unusual one that does not perform data decoding? If it's so, it is correct to utilize `utils.ensure_text_type()` as well in IP fields, since `ipaddress` utilities will not treat `b"192.168.0.1"` the same way as `"192.168.0.1"`. I'll wait for additional feedback on that matter.

Reply:
To be compatible with other field interfaces I've utilised `utils.ensure_text_type()` helper function. From the IP fields perspective it is acceptable - all expected behaviours are consistent with original approach. Numeric (ie. `int(1) -> 0.0.0.1`) and binary (ie. `b"\x00\x00\x00\x01" -> 0.0.0.1`) inputs will not be interpreted by `ip_address` library utilities. 

If the usage of `utils.ensure_text_type()` is obsolete or somehow stems from no longer applicable interface requirements, I belive it shpould be discussed in another issue, and appropriate patch should be applied on all fields utilizing the function.